File: ruby/number_test.rb
require 'minitest/autorun'
require 'bigdecimal'

describe "number" do

  it "plus" do
    assert_equal 5, 3 + 2
  end

  it "minus" do
    assert_equal 1, 3 - 2
  end

  it "divide" do
    assert_equal 1, 3 / 2
  end

  it "multiply" do
    assert_equal 6, 3 * 2
  end

  it "remainder" do
    assert_equal 1, 3 % 2
  end

  it "power" do
    assert_equal 9, 3 ** 2
  end

  it "string to number" do
    assert_equal 1, "1".to_i
    assert_equal 0, nil.to_i
    assert_equal 1.1, "1.1".to_f
  end

  it "abs" do
    assert_equal 10, -10.abs
  end

  it "floor" do
    assert_equal 1, 1.3.floor
  end

  it "round" do
    assert_equal 2, 1.5.round
    assert_equal 1.5, 1.54.round(1)
  end

  it "IEEE 754 float" do
    assert_equal 110.00000000000001, 100 * 1.1
    assert_equal 0.30000000000000004, 0.1 + 0.2
  end

  it "accurate float calculation" do
    assert_equal 110, 100 * BigDecimal("1.1")
    assert_equal 0.3, 0.1 + BigDecimal("0.2")
  end

  it "cos" do
    # http://ruby-doc.org/core-2.2.0/Math.html#cos-method
    assert_equal -1, Math.cos(Math::PI)
  end

end

File: elixir/test/number_test.exs
ExUnit.start

defmodule NumberTest do
  use ExUnit.Case, async: true

  test "plus" do
    assert 3 + 2 = 5
  end

  test "minus" do
    assert 3 - 2 = 1
  end

  test "divide" do
    assert 3 / 2 = 1.5
  end

  test "multiply" do
    assert 3 * 2 = 6
  end

  test "remainder" do
    assert 1 = rem 3, 2
  end

  test "power" do
    assert 9.0 = :math.pow 3, 2
  end

  test "string to number" do
    assert {1, _} = Integer.parse("1")
    assert :error = Integer.parse("")
    assert {1.1, _} = Float.parse("1.1")
  end

  test "abs" do
    assert 10 = abs(-10)
  end

  test "floor" do
    assert 1.0 = Float.floor(1.3)
  end

  test "round" do
    assert 2.0 = Float.round(1.5)
    assert 1.5 = Float.round(1.54, 1)
  end

  test "IEEE 754 float" do
    assert 110.00000000000001 = 100 * 1.1
    assert 0.30000000000000004 = 0.1 + 0.2
  end

  test "accurate float calculation" do
    # no support
  end

  test "cos" do
    assert -1.0 = :math.cos(:math.pi)
  end

end
